File: lab4/c/model_solutions/skiplist.c
// Giles Reger, 2019

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <limits.h>
#include <stdbool.h>

#include "global.h"
#include "pq.h"

// This will have a large impact on performance, try playing with it
#define MAX_LEVEL 20 


struct node {
  int priority;
  Value_Type value;
  // Array of 'forward' pointers
  struct node** next;
  // Just for printing
  int height;
};

// Note that the skiplist is implemented as a linked list of 'nodes'
// where a node represents a tower in the abstract view
struct skiplist {
  int levels;
  int size;
  struct node *header;
};

// Helper function for making nodes, does not initialise next array
struct node* make_node(Value_Type value, int priority, int levels){
  struct node* node = malloc(sizeof(struct node));
  check(node);
  node->priority = priority;
  node->value = value;
  node->next = malloc(sizeof(struct node*) * levels);
  check(node->next);
  node->height=levels;
  return node;
}

struct skiplist* initialize_pq (int size){
  struct skiplist* slist = malloc(sizeof(struct skiplist));
  check(slist);
  slist->levels = 1;
  slist->size = 0; 
  // We will actually use the same sentinel node for start and end 
  slist->header= make_node(NULL,INT_MAX,MAX_LEVEL);
  for(int i=0; i<MAX_LEVEL; i++){
    slist->header->next[i] = slist->header;
  }
  return slist;
}     

void tidy (struct skiplist* list){
  struct node* next = list->header->next[0];
  while(next!=list->header){
    struct node* n = next;
    next = n->next[0];
    free(n->next);
    free(n);
  }
  free(list->header->next);
  free(list->header);
  free(list);
}

bool is_empty(struct skiplist* slist){
  return slist->size <= 0;
}

// Returns the last node with priority less than or equal to 'priority'
// (In the presence of duplicates we could return the first node in the case of equals) 
// Records in 'updates' the nodes along the path that would need updating if a node to
// their right on their level were to be inserted e.g. the nodes at which the decision
// to go 'down' is made
struct node* search(struct skiplist* slist, int priority, struct node** updates)
{
  struct node* node = slist->header;
  int level = MAX_LEVEL;
  while(level >0){
    level--;
    while(node->next[level]->priority <= priority){ 
      node = node->next[level];
    }
    if(updates){updates[level]=node;}
  }
  return node;
}

void insert(struct skiplist* slist, Value_Type value, int priority){

  struct node* updates[MAX_LEVEL];
  struct node* insert_at = search(slist,priority,updates);

  int levels = 1;
  while(levels < MAX_LEVEL && (rand() % 2 == 0)){ levels++;}

  struct node* new_node = make_node(value,priority,levels);

  for(int i=0;i<levels;i++){
    new_node->next[i] = updates[i]->next[i];
    updates[i]->next[i] = new_node;
  }

  slist->size++;
}

bool contains(struct skiplist* slist, Value_Type value, int priority)
{
  struct node* node = search(slist,priority,NULL);
  while(node->priority==priority && node->value && compare(node->value,value)!=0){
    node = node->next[0];
  }
  return (node->priority==priority && node->value && compare(node->value,value)==0);
}


Value_Type pop_min(struct skiplist* slist){

 struct node* min = slist->header->next[0];
 Value_Type res = min->value;

 for(int i=0;i<min->height;i++){
   slist->header->next[i] = min->next[i];
 }
 free(min->next);
 free(min);
 
 slist->size--;

 return res;
}

// There are probably nicer ways to print a skiplist
void print(struct skiplist* list){

  struct node* node = list->header;
  do{
    printf("(%s,%d,%d)\n",node->value,node->priority,node->height);
    node = node->next[0];
  }
  while(node != list->header);
}

File: lab4/python/model_solutions/skiplist.py
import sys
import random

# This will have a large impact on performance, try playing with it
MAX_LEVEL = 20

class node:
    def __init__(self, value, priority,levels):
        self.priority = priority
        self.value = value
        self.height = levels
        self.next = [None] * levels

class skiplist:
    def __init__(self):
        self.levels = 1
        self.size = 0
        self.header = node(None,sys.maxsize,MAX_LEVEL)
        for i in range(MAX_LEVEL):
            self.header.next[i] = self.header
            
    def is_empty(self):
        return self.size <= 0
        
    # Returns the last node with priority less than 'priority'
    #
    # Records in 'updates' the nodes along the path that would need updating if a node to
    # their right on their level were to be inserted e.g. the nodes at which the decision
    # to go 'down' is made
    def search(self, priority, updates):
        node = self.header
        level = MAX_LEVEL
        while (level > 0):
            level = level - 1
            
            while (node.next[level].priority < priority):
                node = node.next[level]
            
            # Record the node where we go down at a particular level
            if (not updates == None):
                updates[level] = node
        
        return node
        
    def insert(self, value, priority):
        updates = [None] * MAX_LEVEL
        insert_at = self.search(priority, updates)
        
        levels = 1
        while (levels < MAX_LEVEL and random.randint(0,1) == 0):
            levels = levels + 1
        
        new_node = node(value,priority,levels)
        
        for i in range(levels):
            new_node.next[i] = updates[i].next[i]
            updates[i].next[i] = new_node
            
        self.size = self.size + 1
        
    def contains(self, value, priority):
        node = self.search(priority, None).next[0]
        while (node.priority == priority and (not node.value == None) and (not node.value == value)):
            node = node.next[0]
        return (node.priority == priority and (not node.value == None) and node.value == value)
        
    def pop_min(self):
        min = self.header.next[0]
        res = min.value
        
        for i in range(min.height):
            self.header.next[i] = min.next[i]
        
        self.size = self.size - 1
        return res
        
    # There are probably nicer ways to print a skiplist
    def print(self):
    
        node = self.header
        sys.stdout.write("(%s,%s,%s)\n" % (node.value, node.priority, node.height))
        node = node.next[0]
        while (not node == self.header):
            sys.stdout.write("(%s,%s,%s)\n" % (node.value, node.priority, node.height))
            node = node.next[0]
        
